% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/dart2df_genind_plink.R
\name{dart2df_genind_plink}
\alias{dart2df_genind_plink}
\title{Swiss Army knife tool to prepare \href{http://www.diversityarrays.com}{DArT} 
output file for population genetics analysis.}
\usage{
dart2df_genind_plink(data, strata, output, pop.levels = NULL,
  blacklist.id = NULL, pop.select = NULL, filter.monomorphic = TRUE,
  common.markers = TRUE, filter.ind.heterozygosity = NULL,
  plot.ind.heterozygosity = FALSE, filter.reproducibility = NULL,
  plot.reproducibility = FALSE, filter.coverage.high = NULL,
  filter.coverage.low = NULL, plot.coverage = FALSE,
  filter.call.rate = NULL, plot.call.rate = FALSE,
  filter.ind.missing.geno = NULL, plot.ind.missing.geno = FALSE,
  filter.markers.missing.ind = NULL, plot.markers.missing.ind = FALSE,
  filter.snp.ld = NULL, plot.number.snp.reads = FALSE,
  maf.thresholds = NULL, maf.pop.num.threshold = 1, maf.operator = "OR",
  filename = NULL, imputation.method = NULL,
  hierarchical.levels = "populations", pred.mean.matching = 0,
  num.tree = 100, verbose = FALSE, parallel.core = detectCores() - 1)
}
\arguments{
\item{data}{DArT output file in wide format or binary format tipically 
used by CSIRO genomic projects.}

\item{strata}{A tab delimited file with columns header:
\code{INDIVIDUALS} and \code{POP_ID}. 
Note: the column \code{POP_ID} refers to any grouping of individuals. If a third column
named \code{NEW_ID} is used, this column will be used to replace the
\code{INDIVIDUALS} column in the main data file.}

\item{output}{Several options: tidy, genind, genlight, vcf, plink, genepop, 
structure. Use a character string,
e.g. \code{output = c("genind", "genepop", "structure")}, to have preferred
output formats generated. The tidy format is generated automatically.}

\item{pop.levels}{(optional string) A character string with your populations ordered.}

\item{blacklist.id}{(optional) A blacklist with individual ID and
a column header 'INDIVIDUALS'. The blacklist is in the working directory
(e.g. "blacklist.txt").
Default: \code{blacklist.id = NULL}.}

\item{pop.select}{(optional string) Keep specific populations. 
Default = \code{NULL} for no selection and keep all population.
e.g. \code{pop.select = c("QUE", "ONT")} to select QUE and ONT population samples.}

\item{filter.monomorphic}{(optional, logical) Should the monomorphic markers present in 
the dataset be filtered out ? Default: \code{filter.monomorphic = TRUE}.}

\item{common.markers}{(optional, logical) Default: \code{common.markers = TRUE}, 
will only keep markers in common (genotyped) between all the populations.}

\item{filter.ind.heterozygosity}{(optional, numerical) Filter the 
individual heterozygosity (averaged accross markers). This argument is useful
to detect mixed samples. The individuals to remove will standout in the plot.
Default: \code{filter.ind.heterozygosity = NULL}.
e.g to remove individuals with heterozygosity higher than 10% use:
\code{filter.ind.heterozygosity = 0.1}.}

\item{plot.ind.heterozygosity}{(optional, logical) Plot the distribution 
(boxplot and jitter plot) of individual heterozygosity. 
Default: \code{plot.ind.heterozygosity = FALSE}.}

\item{filter.reproducibility}{(optional, numerical) Filter the \code{RepAvg} 
column in the data set. Default: \code{filter.reproducibility = NULL}.
e.g to keep markers with reproducibility >= 99%,
use: \code{filter.reproducibility = 0.99}.}

\item{plot.reproducibility}{(optional, logical) Plot the distribution 
(violin plot) of reproducibility. Default: \code{plot.reproducibility = FALSE}.}

\item{filter.coverage.high}{(optional, numerical) Filter the upper bound of the
\code{AvgCountSnp} column in the data set. Default: \code{filter.coverage.high = NULL}.
e.g to keep markers with coverage <= 150 (depth of coverage),
use : \code{filter.coverage.high = 150}.}

\item{filter.coverage.low}{(optional, numerical) Filter the lower bound of the 
\code{AvgCountSnp} column in the data set. Default: \code{filter.coverage.low = NULL}.
e.g to keep markers with coverage >= 10 (depth of coverage),
use : \code{filter.coverage.low = 10}.}

\item{plot.coverage}{(optional, logical) Plot the coverage distribution 
(violin plot). Default: \code{plot.coverage = FALSE}.}

\item{filter.call.rate}{(optional, numerical) Filter the \code{CallRate} 
column in the data set. Default: \code{filter.call.rate = NULL}. e.g to keep 
markers genotyped in more than 95% of the individuals use :
\code{filter.call.rate = 0.95}}

\item{plot.call.rate}{(optional, logical) Plot the distribution 
(violin plot) of call rate. Default: \code{plot.call.rate = FALSE}.}

\item{filter.ind.missing.geno}{(optional, numerical) Filter with the individual's
proportion of missing genotype. Similar to call rate.
e.g to keep individuals genotyped at >= 0.90 of the markers, use:
\code{filter.ind.missing.geno = 0.90}.
Default: \code{filter.ind.missing.geno = NULL}.}

\item{plot.ind.missing.geno}{(optional, logical) Plot the distribution 
(violin plot) of an individual's genotype proportion. 
Default: \code{plot.ind.missing.geno = FALSE}.}

\item{filter.markers.missing.ind}{(optional, numerical) Filter markers based 
on the proportion of individuals genotyped.
e.g to keep markers with >= 0.95 of the genotyped individuals, use:
\code{filter.markers.missing.ind = 0.95}.
Default: \code{filter.markers.missing.ind = NULL}.}

\item{plot.markers.missing.ind}{(optional, logical) Plot the distribution 
(violin plot) of markers genotype individuals proportion. 
Default: \code{plot.markers.missing.ind = FALSE}.}

\item{filter.snp.ld}{(optional, character) With anonymous markers from
reduce representation library like RADseq/GBS/DArT de novo discovery, 
you can explore the impact of the number of snp on the read (haplotype/locus).
This argument is used to minimize linkage disequilibrium (LD) on short reads,
by choosing among these options: 
\code{"1snp"} to keep only reads with 1 SNP/reads,
\code{"2snp"} to keep only reads with at most 2 SNP/reads,
\code{"random"} to select 1 SNP/reads randomly, 
\code{"first"} to select the first SNP on all reads, 
\code{"last"} to select the last SNP on all reads 
(last SNP are usually associated to higher error rate...). 
Default: \code{snp.ld = NULL}.
Note: for long linkage detection use PLINK linkage disequilibrium based SNP 
pruning.}

\item{plot.number.snp.reads}{(optional, logical) Plot the distribution of SNP
per read. 
Default: \code{plot.number.snp.reads = FALSE}.}

\item{maf.thresholds}{(string, double, optional) String with 
local/populations and global/overall maf thresholds, respectively.
Default: \code{maf.thresholds = NULL}. 
e.g. \code{maf.thresholds = c(0.05, 0.1)} for a local maf threshold 
of 0.05 and a global threshold of 0.1.}

\item{maf.pop.num.threshold}{(integer, optional) When maf thresholds are used,
this argument is for the number of pop required to pass the maf thresholds
to keep the locus. Default: \code{maf.pop.num.threshold = 1}}

\item{maf.operator}{(character, optional) \code{maf.operator = "AND"} or default \code{maf.operator = "OR"}.
When filtering over LOCUS or SNP, do you want the local \code{"AND"}
global MAF to pass the thresholds, or ... you want the local \code{"OR"}
global MAF to pass the thresholds, to keep the marker?}

\item{filename}{(optional) The filename prefix for the objet in the global environment 
or the working directory. Default: \code{filename = NULL}. A default name will be used,
customized with the output file(s) selected.}

\item{imputation.method}{(character, optional) 
Methods available for map-independent imputations of missing genotype: 
(1) \code{imputation.method = "max"} to use the most frequent category for 
the imputations.
(2) \code{imputation.method = "rf"} for using Random Forests algorithm.
(3) \code{imputation.method = "boost"} for using extreme gradient boosting trees.
\code{imputation.method = NULL} will return the original dataset, without
imputation.
Default: \code{imputation.method = "rf"}.}

\item{hierarchical.levels}{(character, optional) \code{c("global", "populations")}.
Should the imputations be computed globally or by populations. Note that
imputing genotype globally can create huge bias for example by
introducing foreign genotypes in some populations (see note for more info).
Default: \code{hierarchical.levels = "populations"}.}

\item{pred.mean.matching}{(integer, optional) Used in conjunction with 
random Forests (\code{imputation.method = "rf"}).
Number of candidate non-missing
value to sample from during the predictive mean matching step.
A fast k-nearest neighbor searching algorithms is used with this approach.
\code{pred.mean.matching = 3} will use 3 nighbors.
Default: \code{pred.mean.matching = 0}, avoids this step.}

\item{num.tree}{(integer, optional) The number of trees to grow with the 
when \code{imputation.method = "rf"}.
Default: \code{num.tree = 50}.}

\item{verbose}{(optional, logical) When \code{verbose = TRUE} 
the function is a little more chatty during execution.
Default: \code{verbose = TRUE}.}

\item{parallel.core}{(optional) The number of core used for parallel
execution when \code{imputation.method = "rf"}.
Markers are imputed in parallel, populations are processed sequentially.
Default: \code{parallel::detectCores() - 1}.}
}
\value{
The function returns an object (list). The content of the object 
can be listed with \code{names(object)} and use \code{$} to isolate specific
object (see examples). Some output format will write the output file in the 
working directory. The tidy genomic data frame is generated automatically.
}
\description{
Import, filter and transform a DArT output file to different formats: 
data frame of genotypes, genind object and/or PLINK \code{tped/tfam} format.
}
\examples{
\dontrun{
testing
}
}
\references{
Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, 
Bender D, et al. (2007)
PLINK: a tool set for whole-genome association and population-based linkage 
analyses. 
American Journal of Human Genetics. 81: 559–575. doi:10.1086/519795

Jombart T, Devillard S, Balloux F. (2010)
Discriminant analysis of principal components: a new method for the analysis 
of genetically structured populations.
BMC Genet. 11: 94. doi:10.1186/1471-2156-11-94.

Jombart T, Ahmed I. (2011) adegenet 1.3-1: new tools for the analysis 
of genome-wide SNP data. 
Bioinformatics. 27: 3070-3071. doi:10.1093/bioinformatics/btr521

Ishwaran H. and Kogalur U.B. (2015). Random Forests for Survival,
 Regression and Classification (RF-SRC), R package version 1.6.1.

Ishwaran H. and Kogalur U.B. (2007). Random survival forests
for R. R News 7(2), 25-31.

Ishwaran H., Kogalur U.B., Blackstone E.H. and Lauer M.S. (2008).
Random survival forests. Ann. Appl. Statist. 2(3), 841--860.
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com}
}
