% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/GBS_assignment.R
\name{GBS_assignment}
\alias{GBS_assignment}
\title{Assignment analysis in gsi_sim with GBS data produced by STACKS workflow}
\usage{
GBS_assignment(vcf.file, whitelist.markers = NULL, snp.LD = NULL,
  common.markers = NULL, maf.local.threshold = NULL,
  maf.global.threshold = NULL, maf.pop.num.threshold = 1,
  maf.approach = "SNP", maf.operator = "OR", marker.number = "all",
  blacklist.id = NULL, pop.levels, pop.labels, pop.id.start, pop.id.end,
  pop.select = "all", subsample = NULL, sampling.method, THL = 1,
  iterations = 10, folder, gsi_sim.filename = "gsi_sim_data.txt",
  keep.gsi.files, baseline = NULL, mixture = NULL, imputations = FALSE,
  imputations.group = "populations", num.tree = 100, iteration.rf = 10,
  split.number = 100, verbose = FALSE, parallel.core = NULL)
}
\arguments{
\item{vcf.file}{The VCF file created by STACKS.}

\item{whitelist.markers}{(optional) A whitelist containing CHROM (character 
or integer) and/or LOCUS (integer) and/or 
POS (integer) columns header. To filter by CHROM and/or locus and/or by snp.
The whitelist is in the working directory (e.g. "whitelist.txt"). 
de novo CHROM column with 'un' need to be changed to 1.}

\item{snp.LD}{(optional) Minimize linkage disequilibrium (LD) by choosing 
among these 3 options: \code{"random"} selection, \code{"first"} or 
\code{"last"} SNP on the same read/haplotype. Default = \code{NULL}.}

\item{common.markers}{(optional) Logical. Default = \code{FALSE}. 
With \code{TRUE}, will keep markers present in all the populations.}

\item{maf.local.threshold}{(double) (optional) Filter local/populations maf with a 
threshold before conduncting the assignment analysis. Default = \code{NULL}.}

\item{maf.global.threshold}{(double) (optional) Filter global/overall maf with a 
threshold before conduncting the assignment analysis. Default = \code{NULL}.}

\item{maf.pop.num.threshold}{(integer) When maf threshold is used, 
this argument is for the number of pop required to pass the maf thresholds 
to keep the locus. Default is \code{maf.pop.num.threshold = 1}}

\item{maf.approach}{Character. By \code{maf.approach = "SNP"} or by \code{maf.approach = "haplotype"}. 
The function will consider the SNP or ID/LOCUS/haplotype/read MAF statistics to filter the marker. 
Default is \code{maf.approach = "SNP"}.}

\item{maf.operator}{\code{maf.operator = "AND"} or default \code{maf.operator = "OR"}. 
When filtering over LOCUS or SNP, do you want the local \code{"AND"} 
global MAF to pass the thresholds, or ... you want the local \code{"OR"} 
global MAF to pass the thresholds, to keep the marker?}

\item{marker.number}{(Integer or string of number or "all") Calculations with 
fixed or subsample of your markers. Default= \code{"all"}.
e.g. To test 500, 1000, 2000 and all  the markers: 
\code{marker.number = c(500, 1000, 2000, "all"}. 
To use only 500 makers \code{marker.number = 500}.}

\item{blacklist.id}{(optional) A blacklist with individual ID and
a column header 'INDIVIDUALS'. The blacklist is in the working directory
(e.g. "blacklist.txt").}

\item{pop.levels}{(required) A character string with your populations ordered.}

\item{pop.labels}{(optional) A character string of your populations labels.
If you need to rename sampling sites in \code{pop.levels} or combined sites/pop
into a different names, here is the place.}

\item{pop.id.start}{The start of your population id 
in the name of your individual sample.}

\item{pop.id.end}{The end of your population id 
in the name of your individual sample.}

\item{pop.select}{(string) Conduct the assignment analysis on a 
selected list of populations. Default = \code{"all"}. 
e.g. \code{pop.select = "QUE"} to select QUE population samples. 
\code{pop.select = c("QUE", "ONT")} to select QUE and ONT population samples.}

\item{subsample}{(Integer or Proportion) Default is no sumsampling, \code{subsample = NULL}. 
With a proportion argument \code{subsample = 0.15}, 15 percent of individuals
in each populations are chosen randomly to represent the dataset. 
With \code{subsample = 36}, 36 individuals in each populations are chosen 
randomly to represent the dataset.}

\item{sampling.method}{(character) Should the markers be randomly selected 
\code{"random"} for a classic Leave-One-Out (LOO) assignment or 
chosen based on ranked Fst \code{"ranked"}, used in a 
Traing-Holdout-Leave One Out (THL) assignment ?}

\item{THL}{(character, integer, proportion) For \code{sampling.method = "ranked"} only. 
Default \code{1}, 1 individual sample is used as holdout. This individual is not 
participating in the markers ranking. For each marker number, 
the analysis will be repeated with all the indiviuals in the data set 
(e.g. 500 individuals, 500 times 500, 1000, 2000 markers). 
If a proportion is used e.g. \code{0.15},= 15% of individuals in each 
populations are chosen randomly as holdout individuals. 
With \code{THL = "all"} all individuals are used for ranking (not good) and 
\code{iterations} argument below is set to \code{1} by default. 
For the other THL values, you can create different holdout individuals lists
with the \code{iterations} argument below.}

\item{iterations}{With random marker selection the iterations argument = 
the number of iterations to repeat marker resampling, default is \code{10} 
With \code{marker.number = c(500, 1000)} and default iterations setting, 
500 markers will be randomly chosen 10 times and 1000 markers will be randomly
chosen 10 times. For the ranked method, using \code{THL = 1}, the analysis 
will be repeated for each individuals in the data set for every 
\code{marker.number} selected. With a proportion argument \code{THL = 0.15},
15% of individuals in each populations are chosen randomly as holdout 
individuals and this process is reapeated the number of times chosen by the 
\code{iterations} value.}

\item{folder}{(optional) The name of the folder created in the working directory to save the files/results.}

\item{gsi_sim.filename}{(optional) The name of the file written to the directory.
Use the extension ".txt" at the end. Default \code{gsi_sim_data.txt}.
The number of markers used will be appended to the name of the file.}

\item{keep.gsi.files}{(Boolean) Default \code{FALSE} The input and output gsi_sim files
will be deleted from the directory when finished processing. 
With \code{TRUE}, remember to allocate a large chunk of the disk space for the analysis.}

\item{baseline}{(optional) A character string with your baseline id. 
From the \code{pop.id.start} and \code{pop.id.end} you isolate 
the baseline and mixture group. Here you need to give the id for 
baseline e.g. \code{c("QUE-ADU", "ONT-ADU")}.}

\item{mixture}{(optional) But required if bseline was selected. A character 
string with your mixture id. e.g. \code{c("QUE-JUV", "ONT-JUV")}.}

\item{imputations}{Should a map-independent imputations of markers be
computed. Available choices are: (1) \code{FALSE} for no imputation.
(2) \code{"max"} to use the most frequent category for imputations.
 (3) \code{"rf"} using Random Forest algorithm. Default = \code{FALSE}.}

\item{imputations.group}{\code{"global"} or \code{"populations"}.
Should the imputations be computed globally or by populations. If you choose
global, turn the verbose to \code{TRUE}, to see progress.
Default = \code{"populations"}.}

\item{num.tree}{The number of trees to grow in Random Forest. Default is 100.}

\item{iteration.rf}{The number of iterations of missing data algorithm 
in Random Forest. Default is 10.}

\item{split.number}{Non-negative integer value used to specify 
random splitting in Random Forest. Default is 100.}

\item{verbose}{Logical. Should trace output be enabled on each iteration 
in Random Forest ? Default is \code{FALSE}.}

\item{parallel.core}{(optional) The number of core for OpenMP shared-memory parallel
programming of Random Forest imputations. For more info on how to install the
OpenMP version see \code{\link[randomForestSRC]{randomForestSRC-package}}.
If not selected \code{detectCores()-1} is used as default.}
}
\value{
Depending on arguments selected several files are written to the disk.
In your global environment the results of the assignment analysis.
}
\description{
\code{gsi_sim} is a tool for doing and simulating genetic stock 
identification and developed by Eric C. Anderson.
The arguments in the \code{GBS_assignment} function were tailored for the 
reality of GBS data for assignment analysis while 
maintaining a reproducible workflow. 
The input data is a VCF file produced by STACKS. Individuals, populations and 
markers can be filtered and/or selected in several ways using blacklist, 
whitelist and other arguments. Map-independent imputation of missing genotype 
using Random Forest or the most frequent category is also available. 
Markers can be randomly selected for a classic LOO (Leave-One-Out) 
assignment or chosen based on ranked Fst for a THL 
(Traing, Holdout, Leave-one-out) assignment analysis.
}
\details{
The imputations using Random Forest requires more time to compute 
and can take several
minutes and hours depending on the size of the dataset and polymorphism of
the species used. e.g. with a low polymorphic taxa, and a data set 
containing 30\% missing data, 5 000 haplotypes loci and 500 individuals 
will require 15 min.
The Fst is based on Weir and Cockerham 1984 equations.
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com}
}
\references{
Anderson, Eric C., Robin S. Waples, and Steven T. Kalinowski. (2008) 
An improved method for predicting the accuracy of genetic stock identification.
Canadian Journal of Fisheries and Aquatic Sciences 65, 7:1475-1486.

Anderson, E. C. (2010) Assessing the power of informative subsets of 
loci for population assignment: standard methods are upwardly biased.
Molecular ecology resources 10, 4:701-710.

Catchen JM, Amores A, Hohenlohe PA et al. (2011) 
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences. 
G3, 1, 171-182.

Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013) 
Stacks: an analysis tool set for population genomics. 
Molecular Ecology, 22, 3124-3140.

Weir BS, Cockerham CC (1984) Estimating F-Statistics for the 
Analysis of Population Structure. Evolution, 38, 1358–1370.

Ishwaran H. and Kogalur U.B. (2015). Random Forests for Survival,
 Regression and Classification (RF-SRC), R package version 1.6.1.

Ishwaran H. and Kogalur U.B. (2007). Random survival forests
for R. R News 7(2), 25-31.

Ishwaran H., Kogalur U.B., Blackstone E.H. and Lauer M.S. (2008).
Random survival forests. Ann. Appl. Statist. 2(3), 841--860.
}

