% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/stackr_maf_module.R
\name{stackr_maf_module}
\alias{stackr_maf_module}
\title{MAF computation from a tidy data frame}
\usage{
stackr_maf_module(data, maf.thresholds = NULL, maf.pop.num.threshold = 1,
  maf.approach = "SNP", maf.operator = "OR")
}
\arguments{
\item{data}{A biallelic genomic data set in the working directory or 
object in the global environment in wide or long (tidy) formats. 
\emph{How to get a tidy data frame ?}
Look into \pkg{stackr} \code{\link{tidy_genomic_data}}.}

\item{maf.thresholds}{(string, double, optional) String with 
local/populations and global/overall maf thresholds, respectively.
e.g. \code{maf.thresholds = c(0.05, 0.1)} for a local maf threshold 
of 0.05 and a global threshold of 0.1. Available for VCF, PLINK and data frame 
files.
Default: \code{maf.thresholds = NULL}.}

\item{maf.pop.num.threshold}{(integer, optional) When maf thresholds are used,
this argument is for the number of pop required to pass the maf thresholds
to keep the locus. Default: \code{maf.pop.num.threshold = 1}}

\item{maf.approach}{(character, optional). 
\code{maf.approach = "haplotype"} : looks at the minimum MAF found on the 
read/haplotype. Using this option will discard all the markers/snp on 
that read based on the thresholds chosen. This method is only available 
for VCF and haplotype files, or tidy data frame from those file types.
\code{maf.approach = "SNP"} : treats all the SNP on the same 
haplotype/read as independent. Doesn't work with haplotype file, 
but does work for all other file type.
Default is \code{maf.approach = "SNP"}.}

\item{maf.operator}{(character, optional) \code{maf.operator = "AND"} or 
default \code{maf.operator = "OR"}.
When filtering over LOCUS or SNP, do you want the local \code{"AND"}
global MAF to pass the thresholds, or ... you want the local \code{"OR"}
global MAF to pass the thresholds, to keep the marker?}
}
\value{
A tidy data frame with 6 columns: 
\code{MARKERS, INDIVIDUALS, REF, ALT, GT_VCF, GT_BIN}.
\code{GT_VCF}: the genotype in VCF format
\code{GT_BIN}: coding used internally to easily convert to genlight, 
the coding \code{0, 1, 2, NA} stands for the number of ALT allele in the 
genotype and \code{NA} for missing genotype.
}
\description{
Compute the minor allele frequency (local and global) using 
a genomic tidy data frame.
Used internally in \href{https://github.com/thierrygosselin/stackr}{stackr} 
and might be of interest for users.
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com}
}

