% Generated by roxygen2 (4.1.1): do not edit by hand
% Please edit documentation in R/summary.R
\name{summary_haplotypes}
\alias{summary_haplotypes}
\title{Haplotypes file summary v2}
\usage{
summary_haplotypes(haplotypes.file, whitelist.loci = NULL,
  blacklist.id = NULL, pop.id.start, pop.id.end, pop.levels, read.length)
}
\arguments{
\item{haplotypes.file}{The 'batch_x.haplotypes.tsv' created by STACKS.}

\item{whitelist.loci}{(optional) A whitelist of loci and
a column header 'LOCUS'.
The whitelist is in the directory (e.g. "whitelist.txt").}

\item{blacklist.id}{(optional) A blacklist with individual ID and
a column header 'INDIVIDUALS'. The blacklist is in the directory
 (e.g. "blacklist.txt").}

\item{pop.id.start}{The start of your population id
in the name of your individual sample.}

\item{pop.id.end}{The end of your population id
in the name of your individual sample.}

\item{pop.levels}{An optional character string with your populations ordered.}

\item{read.length}{The length in nucleotide of your reads (e.g. 80 or 100).}
}
\value{
The function returns a list with the summary, the paralogs and
consensus loci by populations and unique loci (use $ to access each
components).
Write 3 files in the working directory:
blacklist of unique putative paralogs and unique consensus loci
and a summary of the haplotypes file by population.
}
\description{
STACKS batch_x.haplotypes.tsv file summary.
Output summary table for populations with putative paralogs,
consensus, monomorphic and polymorphic loci.
The haplotypes statistics for the observed and expected homozygosity and
heterozygosity. Wright’s inbreeding coefficient (Fis), and a proxy measure of
the realized proportion of the genome that is identical by descent (IBDG),
the FH measure based on the excess in the observed number of homozygous
genotypes within an individual relative to the mean number of homozygous
genotypes expected under random mating
(Keller et al., 2011; Kardos et al., 2015).
The nucleotide diversity (Pi) is also given. Pi measured here consider the
consensus loci in the catalog (no variation between population sequences).
}
\details{
haplo.summary <- haplotype.file.summary$summary

paralogs.pop <- haplotype.file.summary$paralogs.pop

paralogs.loci <- haplotype.file.summary$paralogs.loci

consensus.pop <- haplotype.file.summary$consensus.pop

consensus.loci <- haplotype.file.summary$consensus.loci
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com} and
Anne-Laure Ferchaud \email{annelaureferchaud@gmail.com}
}
\references{
Keller MC, Visscher PM, Goddard ME (2011)
Quantification of inbreeding due to distant ancestors and its detection
 using dense single nucleotide polymorphism data. Genetics, 189, 237–249.

Kardos M, Luikart G, Allendorf FW (2015)
Measuring individual inbreeding in the age of genomics: marker-based
measures are better than pedigrees. Heredity, 115, 63–72.

Nei M, Li WH (1979)
Mathematical model for studying genetic variation in terms of
restriction endonucleases.
Proceedings of the National Academy of Sciences of
the United States of America, 76, 5269–5273.
#'
}

