% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/vcf2plink.R
\name{vcf2plink}
\alias{vcf2plink}
\title{VCF to plink with filters and data imputation}
\arguments{
\item{data}{The VCF file created by STACKS.}

\item{whitelist.markers}{(optional) A whitelist containing CHROM (character or integer) and/or LOCUS (integer) and/or 
POS (integer) columns header. To filter by CHROM and/or locus and/or by snp.
The whitelist is in the directory (e.g. "whitelist.txt"). de novo CHROM column with 'un' need to be changed to 1.}

\item{blacklist.genotype}{(optional) Useful to erase genotype with below 
average quality, e.g. genotype with more than 2 alleles in diploid likely 
sequencing errors or genotypes with poor genotype likelihood or coverage. 
The blacklist as a minimum of 2 column headers (markers and individuals). 
Markers can be 1 column (CHROM or LOCUS or POS), 
a combination of 2 (e.g. CHROM and POS or CHROM and LOCUS or LOCUS and POS) or 
all 3 (CHROM, LOCUS, POS) The markers columns must be designated: CHROM (character
or integer) and/or LOCUS (integer) and/or POS (integer). The id column designated
INDIVIDUALS (character) columns header. The blacklist must be in the working 
directory (e.g. "blacklist.genotype.txt"). For de novo VCF, CHROM column 
with 'un' need to be changed to 1. Default \code{NULL} for no blacklist of 
genotypes to erase.}

\item{snp.ld}{(optional) Minimize linkage disequilibrium (LD) by choosing 
among these 3 options: \code{"random"} selection, \code{"first"} or 
\code{"last"} SNP on the same read/haplotype. Default = \code{NULL}.}

\item{common.markers}{(optional) Logical. Default = \code{FALSE}. 
With \code{TRUE}, will keep markers present in all the populations.}

\item{blacklist.id}{(optional) A blacklist with individual ID and
a column header 'INDIVIDUALS'. The blacklist is in the directory
 (e.g. "blacklist.txt").}

\item{pop.levels}{(required) A character string with your populations ordered.}

\item{pop.labels}{(optional) A character string of your populations labels.
If you need to rename sampling sites in \code{pop.levels} or combined sites/pop
into a different names, here is the place.}

\item{pop.id.start}{The start of your population id 
in the name of your individual sample.}

\item{pop.id.end}{The end of your population id 
in the name of your individual sample.}

\item{strata}{(optional) A tab delimited file with 2 columns with header:
\code{INDIVIDUALS} and \code{STRATA}. Default: \code{strata = NULL}. 
This is especially useful if you have a hierarchical or factorial sampling 
design. See also \code{\link[adegenet]{genind}} for details.}

\item{pop.select}{(string) Conduct the conversion on a
selected list of populations. Default = \code{NULL} for no selection and keep
all population.
e.g. \code{pop.select = "QUE"} to select QUE population samples.
\code{pop.select = c("QUE", "ONT")} to select QUE and ONT population samples.}

\item{imputation.method}{(character, optional) 
Methods available for map-independent imputations of missing genotype: 
(1) \code{"max"} to use the most frequent category for imputations.
(2) \code{"rf"} using Random Forest algorithm. 
Default: no imputation \code{imputation.method = NULL}.}

\item{impute}{(character, optional) Imputation on missing genotype 
\code{impute = "genotype"} or alleles \code{impute = "allele"}.
Default: \code{"genotype"}.}

\item{imputations.group}{(character, optional) \code{"global"} or \code{"populations"}.
Should the imputations be computed globally or by populations. If you choose
global, turn the verbose to \code{TRUE}, to see progress.
Default = \code{"populations"}.}

\item{num.tree}{(integer, optional) The number of trees to grow in Random Forest. 
Default: \code{num.tree = 100}.}

\item{iteration.rf}{(integer, optional) The number of iterations of missing data algorithm
in Random Forest. 
Default: \code{iteration.rf = 10}.}

\item{split.number}{(integer, optional) Non-negative integer value used to specify
random splitting in Random Forest. 
Default: \code{split.number = 100}.}

\item{verbose}{(logical, optional) Should trace output be enabled on each iteration
in Random Forest ? 
Default: \code{verbose = FALSE}.}

\item{parallel.core}{(optional) The number of core for OpenMP shared-memory parallel
programming of Random Forest imputations. For more info on how to install the
OpenMP version see \code{\link[randomForestSRC]{randomForestSRC-package}}.
If not selected \code{detectCores()-1} is used as default.}

\item{filename}{(optional) The prefix for the PLINK file written to the directory.
"tped" and "tfam" will be appended by the function. 
Default \code{data_date@time.tped, data_date@time.tped}.}
}
\value{
When no imputation is selected the PLINK files are saved to the 
working directory. When imputation is selected 4 files are saved to
the working directory. The PLINK file returned is in the \code{tped/tfam} 
format. The \code{strata} or the population arguments above are used 
for first 2 columns of the \code{tfam} file.
}
\description{
This function can first filter the vcf file 
with a whitelist of loci and a blacklist of individuals (optional). 
Then it will convert the file
to a PLINK tped and tfam file.
Map-independent imputation using Random Forest or the most frequent category
is also available as an option.
}
\details{
The imputations using Random Forest requires more time to compute and can take several
minutes and hours depending on the size of the dataset and polymorphism of
the species used. e.g. with a low polymorphic taxa, and a data set 
containing 30\% missing data, 5 000 haplotypes loci and 500 individuals 
will require 15 min.
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com}
}
\references{
Catchen JM, Amores A, Hohenlohe PA et al. (2011) 
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences. 
G3, 1, 171-182.

Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013) 
Stacks: an analysis tool set for population genomics. 
Molecular Ecology, 22, 3124-3140.

Ishwaran H. and Kogalur U.B. (2015). Random Forests for Survival,
 Regression and Classification (RF-SRC), R package version 1.6.1.

Ishwaran H. and Kogalur U.B. (2007). Random survival forests
for R. R News 7(2), 25-31.

Ishwaran H., Kogalur U.B., Blackstone E.H. and Lauer M.S. (2008).
Random survival forests. Ann. Appl. Statist. 2(3), 841-860.

Danecek P, Auton A, Abecasis G et al. (2011)
The variant call format and VCFtools.
Bioinformatics, 27, 2156-2158.

Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, 
Bender D, et al. 
PLINK: a tool set for whole-genome association and population-based linkage 
analyses. 
American Journal of Human Genetics. 2007; 81: 559â€“575. doi:10.1086/519795
}
\seealso{
\code{randomForestSRC} is available on CRAN \url{http://cran.r-project.org/web/packages/randomForestSRC/} and github \url{https://github.com/ehrlinger/randomForestSRC}
}

