---
output:
  md_document:
    variant: markdown_github
---
[![Travis-CI Build Status](https://travis-ci.org/thierrygosselin/stackr.svg?branch=master)](https://travis-ci.org/thierrygosselin/stackr)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/thierrygosselin/stackr?branch=master&svg=true)](https://ci.appveyor.com/project/thierrygosselin/stackr)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/stackr)](http://cran.r-project.org/package=stackr)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![DOI](https://zenodo.org/badge/14548/thierrygosselin/stackr.svg)](https://zenodo.org/badge/latestdoi/14548/thierrygosselin/stackr)

```{r, echo = FALSE}
description <- readLines("DESCRIPTION")
rvers <- stringr::str_match(grep("R \\(", description, value = TRUE), "[0-9]{1,4}\\.[0-9]{1,4}\\.[0-9]{1,4}")[1,1]
```
```{r, echo = FALSE}
description <- readLines("DESCRIPTION")
version <- gsub(" ", "", gsub("Version:", "", grep("Version:", description, value = TRUE)))
```
[![packageversion](https://img.shields.io/badge/Package%20version-`r version`-orange.svg)](commits/master)
[![Last-changedate](https://img.shields.io/badge/last%20change-`r gsub('-', '--', Sys.Date())`-brightgreen.svg)](/commits/master)

---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```


# stackr: an R package for the analysis of GBS/RADseq data

This is the development page of the **stackr**, 
if you want to help, see [contributions section](https://github.com/thierrygosselin/stackr#contributions)

Most genomic analysis look for patterns and trends with various statistics. 
Bias, noise and outliers can have bounded influence on estimators and interfere
with polymorphism discovery. Avoid bad data exploration and control the impact
of filters on your downstream genetic analysis. Use stackr to: import, explore,
manipulate, visualize, filter, impute and export your GBS/RADseq data.


## Installation
To try out the dev version of **stackr**, copy/paste the code below:

```r
if (!require("devtools")) install.packages("devtools") # to install
source("https://bioconductor.org/biocLite.R") # for bioconductor dependencies
biocLite() # to install bioconductor
devtools::install_github("thierrygosselin/stackr")
library(stackr) # to load
```



| Caracteristics | Description |
|:-------------------|:--------------------------------------------------------|
| **Import** | List of the 11 supported genomic file formats in `tidy_genomic_format` and `genomic_converter`:<br> [VCF, SNPs and haplotypes](https://samtools.github.io/hts-specs/) (Danecek et al., 2011)<br>[PLINK tped/tfam](http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml#tr) (Purcell et al., 2007)<br>[genind](https://github.com/thibautjombart/adegenet) (Jombart et al., 2010; Jombart and Ahmed, 2011)<br>  [genlight](https://github.com/thibautjombart/adegenet) (Jombart et al., 2010; Jombart and Ahmed, 2011), also in `tidy_genlight`<br>[strataG gtypes](https://github.com/EricArcher/strataG) (Archer et al., 2016)<br>[Genepop](http://genepop.curtin.edu.au) (Raymond and Rousset, 1995; Rousset, 2008), also in `tidy_genepop`<br>[STACKS haplotype file](http://catchenlab.life.illinois.edu/stacks/) (Catchen et al., 2011, 2013)<br> [hierfstat](https://github.com/jgx65/hierfstat) (Goudet, 2005), also in `tidy_fstat`<br>Dataframes of genotypes in wide or long/tidy format, also in `tidy_wide`|
| **Output** |16 genomic data formats can be exported out of **stackr** using `genomic_converter` or these separate modules:<br>`write_vcf`: [VCF](https://samtools.github.io/hts-specs/) (Danecek et al., 2011)<br>`write_plink`: [PLINK tped/tfam](http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml#tr) (Purcell et al., 2007)<br>`write_genind`: [adegenet genind and genlight](https://github.com/thibautjombart/adegenet) (Jombart et al., 2010; Jombart and Ahmed, 2011)<br>`write_genlight`: [genlight](https://github.com/thibautjombart/adegenet) (Jombart et al., 2010; Jombart and Ahmed, 2011)<br>`write_gtypes`: [strataG gtypes](https://github.com/EricArcher/strataG) (Archer et al. 2016)<br>`write_colony`: [COLONY](https://www.zsl.org/science/software/colony) (Jones and Wang, 2010; Wang, 2012)<br>`write_genepop`: [Genepop](http://genepop.curtin.edu.au) (Raymond and Rousset, 1995; Rousset, 2008)<br>[STACKS haplotype file](http://catchenlab.life.illinois.edu/stacks/) (Catchen et al., 2011, 2013)<br>`write_betadiv`: [betadiv](http://adn.biol.umontreal.ca/~numericalecology/Rcode/) (Lamy, 2015)<br> `vcf2dadi`: [δaδi](http://gutengroup.mcb.arizona.edu/software/) (Gutenkunst et al., 2009)<br> `write_structure`: [structure](http://pritchardlab.stanford.edu/structure.html) (Pritchard et al., 2000)<br> `write_arlequin`: [Arlequin](http://cmpg.unibe.ch/software/arlequin35/) (Excoffier et al. 2005)<br> `write_hierfstat`: [hierfstat](https://github.com/jgx65/hierfstat) (Goudet, 2005)<br> `write_snprelate`: [SNPRelate](https://github.com/zhengxwen/SNPRelate) (Zheng et al. 2012)<br>Dataframes of genotypes in wide or long/tidy format|
|**Conversion function**| `genomic_converter` import/export genomic formats mentioned above. The function is also integrated with usefull filters, blacklist and whitelist.|
|**Outliers detection**|`detect_duplicate_genomes`: Detect and remove duplicate individuals from your dataset <br>`detect_mixed_genomes`: Detect and remove potentially mixed individuals<br>`summary_haplotype` and `filter_snp_number`: Discard of outlier markers with *de novo* assembly artifact (e.g. markers with an extreme number of SNP per haplotype or with irregular number of alleles)|
|**Pattern of missingness**|`missing_visualization`: Visualize patterns of missing data. Find patterns associated with different variables of your study (lanes, chips, sequencers, populations, sample sites, reads/samples, homozygosity, etc)|
|**Filters**| Alleles, genotypes, markers, individuals and populations can be filtered and/or selected in several ways:<br><br>`filter_coverage`: Discard markers based on read depth (coverage) of alleles and genotypes<br>`filter_genotype_likelihood`: Discard markers based on genotype likelihood<br>`filter_individual`: Discard markers based on proportion of genotyped individuals<br>`filter_population`: Discard markers based on proportion of genotyped populations<br>`filter_het`: Discard markers based on the observed heterozygosity (Het obs)<br>`filter_fis`: Discard markers based on the inbreeding coefficient (Fis)<br>`filter_hw`: Discard markers based on the Hardy-Weinberg Equilibrium expectations (HWE)<br>`filter_maf`: Discard markers based on local or global (overall) minor allele frequency|
|**Imputations**|**Map-independent** imputations of missing genotypes.<br>Using **Random Forest** or the most frequent category.<br> Imputations can be conducted **overall samples** or **by populations**.<br><br>Imputations are integrated in several functions and in a separate module: `stackr_imputations_module`|
|**[ggplot2](http://ggplot2.org)-based plotting**|Visualize distribution of important metric and statistics and create publication-ready figures|
|**Parallel**|Codes designed and optimized for fast computations running imputations, iterations, etc. in parallel. Works with all OS: Linux, Mac and now PC!|

[More in stackr workflow below](https://github.com/thierrygosselin/stackr#stackr-workflow)



## Prerequisite - Suggestions - Troubleshooting
  * **Parallel computing**: Follow the steps in this [vignette](https://github.com/thierrygosselin/stackr/blob/master/vignettes/vignette_imputations_parallel.Rmd)  to install [data.table](https://github.com/Rdatatable/data.table) and [XGBoost](https://github.com/dmlc/xgboost) packages (e.g. to do imputations in parallel).
  * **Installation problem:** see this
  [vignette](https://github.com/thierrygosselin/stackr/blob/master/vignettes/vignette_installation_problems.Rmd)
  * **Windows users**: Install [Rtools](https://cran.r-project.org/bin/windows/Rtools/).
  * For a better experience in **stackr** and in R in general, I recommend using [RStudio](https://www.rstudio.com/products/rstudio/download/). 
  The R GUI is unstable with functions using parallel ([more info](https://stat.ethz.ch/R-manual/R-devel/library/parallel/html/mclapply.html)). 

## Vignettes, R Notebooks and examples

**Vignettes (in development, check periodically for updates):**

* Vignettes with real data for example in the form of R Notebooks take too much
space to be included in package, without CRAN complaining.
Consequently, vignettes are gradually being excluded from the package and
distributed separately, follow the links below.
* [installation problems](https://github.com/thierrygosselin/stackr/blob/master/vignettes/vignette_installation_problems.Rmd)
* [parallel computing during imputations](https://github.com/thierrygosselin/stackr/blob/master/vignettes/vignette_imputations_parallel.Rmd) 
* [vcf2dadi](https://github.com/thierrygosselin/stackr/blob/master/vignettes/vignette_vcf2dadi.Rmd)
* [haplo2genind](https://github.com/thierrygosselin/stackr/blob/master/vignettes/vignette_haplo2genind.Rmd)

**R Notebooks:**

* Missing data visualization and analysis [(html vignette)](https://www.dropbox.com/s/4zf032g6yjatj0a/vignette_missing_data_analysis.nb.html?dl=0) and [(Rmd)](https://www.dropbox.com/s/5fxw2h9w1l1j391/vignette_missing_data_analysis.Rmd?dl=0)


## Citation:
To get the citation, inside R:
```r
citation("stackr")
```

## New features
Change log, version, new features and bug history now lives in the [NEWS.md file](https://github.com/thierrygosselin/stackr/blob/master/NEWS.md)

**v.0.5.0**

* better parallelization of `summary_haplotypes` function. With progress bar...
* bug fix with `summary_haplotypes` not properly summarizing info when no assembly artifacts were found

**v.0.4.9**

* bug fix with detection of mixed bi/multi allelic dataset. The bug was detected
in `tidy_genomic_data` and `genomic_converter` functions.


**v.0.4.8**

* safer use and better parsing of `strataG` object to work with tidy data and pass Travis CI.


**v.0.4.7**

* Better parsing of genepop file with 2 characters for allele coding
* Only 10% of markers are now used for increase speed of bi-allelic markers detection
* Work on imputation module that will be functional in the next version bump
* Better code for the progress bar (Linux and Mac only) that now shows an ETA along the progress.


**v.0.4.6**

* I'm pleased to announce that `stackr` parallel mode now works with **Windows**!
Nothing to install, just need to choose the number of CPU,
the rest is done automatically.
* `haplo2colony` is deprecated. Use the new function called `write_colony`!
* `write_colony`: works similarly to the deprecated function `haplo2colony`,
      * with the major advantage that it's no longer restricted to STACKS 
      haplotypes file. 
      * The function is using the `tidy_genomic_data` module
      to import files. So you can choose one of the 10 input file formats supported
      by `stackr`!
      * other benefits also include the possibility to efficiently test MAF, 
      snp.ld, haplotypes/snp approach, whitelist of markes, 
      blacklist of individuals, blacklist of genotypes, etc. with the buit-in
      arguments.
      * the function only **keeps markers in common** between populations/groups
      and **is removing monomorphic markers**.
      * **Note:** there are several *defaults* in the function and
      it's a complicated file format... please, read the function
      documentation and `COLONY` manual.


**v.0.4.5**

* temporary fix to `tidy_genomic_data` to read unconventional Tassel VCF

* new function `ibdg_fh` computes the FH measure that was previously
computed in `summary_haplotypes`.
It now works with biallelic and multiallelic data.

    The FH measure is based on the excess in the observed number of homozygous
genotypes within an individual relative to the mean number of homozygous
genotypes expected under random mating (see function for details). The `IBDg` in
the name is because the measure is a proxy of the realized proportion of the genome
that is identical by descent by reference to the current population 
under hypothetical random mating.

* `missing_visualization` now computes the FH measure and look for correlation
with average missingness per individual.
* `tidy_stacks_haplotypes_vcf` is now deprecated in favor of using `tidy_genomic_data`
that will import haplotypic vcf files.

**v.0.4.4**

* several updates to make functions faster.
* `stackr_imputations_module` no longer imputes globally after imputations
by populations. Instead, use `common.markers` or not to test impacts.
*  bug fix with `change_alleles` that was not working properly inside
the imputation module.
* `snp_ld` is not a separate module available for users. Check documentation.
* `missing_visualization` now show the proportion of variance with plot axis text.

For previous news:
[NEWS.md file](https://github.com/thierrygosselin/stackr/blob/master/NEWS.md)

## Roadmap of future developments:

* Integrate more imputation method into `stackr_imputations_module` (`missRanger`, kNNi-LD approaches).
* Updated filters: more efficient, interactive and visualization included: *in progress*.
* Workflow tutorial that links functions and points to specific vignettes to further explore some problems: *in progress*
* Integration of several functions with [STACKS](http://catchenlab.life.illinois.edu/stacks/) and [DArT](http://www.diversityarrays.com) database *in progress*.
* Use Shiny and ggvis (when subplots and/or facets becomes available for ggvis).
* Until publication **stackr** will change rapidly, stay updated with releases and contribute with bug reports.
* Suggestions ?


## Contributions:

This package has been developed in the open, and it wouldn’t be nearly as good without your contributions. There are a number of ways you can help me make this package even better:

* If you don’t understand something, please let me know. 
* Your feedback on what is confusing or hard to understand is valuable. 
* If you spot a typo, feel free to edit the underlying page and send a pull request.

New to pull request on github ? The process is very easy:

* Click the edit this page on the sidebar.
* Make the changes using github’s in-page editor and save.
* Submit a pull request and include a brief description of your changes. 
* “Fixing typos” is perfectly adequate.


## GBS workflow
The **stackr** package fits currently at the end of the GBS workflow. Below, a flow chart using [STACKS](http://catchenlab.life.illinois.edu/stacks/) and other software.
![](vignettes/GBS_workflow.png)

## stackr workflow 
Currently under construction. Come back soon!

**Table 1: Quality control and filtering RAD/GBS data**

| Parameters | Libraries & Seq.Lanes | Alleles | Genotypes | Individuals | Markers | Sampling sites | Populations | Globally |
|:----|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|
| Quality |x| | |x| | | | |
| Assembly and genotyping |x| | | | | | | |
| Outliers | |x|x|x|x| | | |
| Missingness |x|x|x|x|x|x|x|x|
| Coverage | |x|x| |x| | | |
| Genotype Likelihood | | |x| | | | | |
| Prop. Genotyped | | | |x|x|x|x|x|
| HET & FIS & HWE | | | |x|x| |x| |
| MAF | | | | |x|x|x|x|
| Missingness |x|x|x|x|x|x|x|x|


**Step 1 Quality** Ask yourself these questions: 

* DNA quality, libraries quality, sequencing lanes quality ? 
* Please stop thinking in terms of quantity (e.g. millions of reads returned), and start thinking about actual quality of your new data.
* Use quality metrics inside available software (e.g. fastqc)

**Step 2 *de novo* assembly and genotyping** 

* This is conducted outside stackr
* Integrated software pipelines include: [STACKS](http://catchenlab.life.illinois.edu/stacks/), [pyRAD](http://dereneaton.com/software/), [dDocent](https://ddocent.wordpress.com), [AftrRAD](http://u.osu.edu/sovic.1/downloads/). If you want to develop your own pipeline, 
there are a multitude of approaches, good luck. 

**Step 3 Outliers**

* Remove replicates (I hope you have some).
* Remove *de novo* assembly artifact, by creating blacklist of genotypes or whitelist of markers:
    * individuals with more than 2 alleles (use `summary_haplotypes`)
    * outlier markers with extreme number of SNP/read or haplotype (use `filter_snp_number`)
* Remove potential duplicated samples that went off your radar, try `detect_duplicate_genomes`.
* Highlight outliers individual's heterozygosity that might represent mixed samples with `detect_mixed_individuals`.
* The metric you're using: a *de novo* artefact or a reliable signal of biological polymorphism?
* Should the statistic you are interested in be consistent throughout the read ?
* Will you consider haplotype or snp level statistics?
* The consistensies of SNPs statistics among haplotype can be throughly tested by using `snp.ld` argument in several stackr functions.
* Any other outliers with different individual's or markers metrics (reads/sample, etc) ?

**Step 4 Pattern of missingness**

* Use `missing_visualization` with/without your new blacklists (e.g. of genotypes, individuals) and with/without whitelist of markers to examine patterns of missingness in you dataset before more extensive filtering (there is a vignette for this step)
* The trick here is to use the `strata` argument to find patterns associated with different variables of your study (lanes, chips, sequencers, populations, sample sites, reads/samples, etc).
* Do you see a trend between your missing pattern and reads/samples ? Heterozygosity?
* Do you need more sequencing? Do you have to re-run some lanes? 

**Step 5-6 Coverage and Genotype Likelihood**

* Coverage is an individual metric. With most software you'll find allele and genotype coverage info.
* Genotype likelihood is usually a metric based on coverage of the different genotypes found in all of your data.
* Good allele coverage is required for reliable genotypes.
* Reliable genotypes is required for reliable downstream summary statistics.
* Explore filtering options in `filter_coverage` and `filter_genotype_likelihood`.

**Step 7 Prop. Genotyped**

* Do you have enough individuals in each sampling sites (`filter_individual`) 
and enough putative populations (`filter_population`) for each markers ?
* Use blacklist of individuals with different thresholds.
* Keep different whitelist of markers.
* Use `common.markers` argument inside most of stackr functions to test the impact of vetting loci based on shared markers.
* Use imputation methods provided by stackr (inside `tidy_genomic_data` or `genomic_converter`, as a separate module: `stackr_imputations_module`) to assess the impact of lowering or increasing threshold that impact missing data.

**Step 8 HET, Fis, HWE**

* Overall and/or per populations hwe, heterozygosity and Fis statistics can highlight: 
*de novo* assembly problems (oversplitting/undermerging), genotyping problems or
biological problems.
* These filters allows to test rapidly if departure from realistic expectations
are a problem for downstream analysis ?
* Choose your threshold wisely and test impact on pipeline.
* Use `filter_het`, `filter_fis`, `filter_hwe` and look again 
at the individual's heterozygosity (`detect_mixed_individuals`) for outliers.

**Step 9 MAF**

* Remove artifactual and uninformative markers.
* Use MAF arguments inside several of stackr functions to tailor MAF to your analysis tolerance to minor allelel frequencies.
* There is also a separate filter in stackr: `filter_maf`

**Step 10 Pattern of missingness, again**

* Use `missing_visualization` with your new blacklists (e.g. of genotypes, individuals) and with your whitelist of markers to examine patterns of missingness in your dataset after filtering (there is a vignette for this step)
* The trick here is to use the `strata` argument to find patterns associated with different variables of your study (lanes, chips, sequencers, populations, sample sites, reads/samples, etc).
* Do you see a trend between your missing pattern and reads/samples ? Heterozygosity?
