# write a betadiv from a tidy data frame

#' @name write_betadiv

#' @title Write a betadiv file from a tidy data frame

#' @description Write a betadiv file from a tidy data frame.
#' Used internally in \href{https://github.com/thierrygosselin/stackr}{stackr} 
#' and \href{https://github.com/thierrygosselin/assigner}{assigner}
#' and might be of interest for users.
 
#' @param data A tidy data frame object in the global environment or
#' a tidy data frame in wide or long format in the working directory.
#' \emph{How to get a tidy data frame ?}
#' Look into \pkg{stackr} \code{\link{tidy_genomic_data}}.

#' @return A betadiv object is returned.

#' @export
#' @rdname write_betadiv
#' @importFrom dplyr select distinct n_distinct group_by ungroup rename arrange tally filter if_else mutate summarise left_join inner_join right_join anti_join semi_join full_join
#' @importFrom stringi stri_replace_all_fixed
#' @importFrom tidyr spread



#' @references Lamy T, Legendre P, Chancerelle Y, Siu G, Claudet J (2015) 
#' Understanding the Spatio-Temporal Response of Coral Reef Fish Communities to 
#' Natural Disturbances: Insights from Beta-Diversity Decomposition. 
#' PLoS ONE, 10, e0138696.

#' @seealso \code{beta.div} is available on Pierre Legendre web site \url{http://adn.biol.umontreal.ca/~numericalecology/Rcode/} \code{randomForestSRC} is available on CRAN \url{http://cran.r-project.org/web/packages/randomForestSRC/} and github \url{https://github.com/ehrlinger/randomForestSRC}

#' @author Laura Benestan \email{laura.benestan@@icloud.com} and
#' Thierry Gosselin \email{thierrygosselin@@icloud.com}


write_betadiv <- function(data) {
  
  # Import data ---------------------------------------------------------------
  if (is.vector(data)) {
    input <- stackr::tidy_wide(data = data, import.metadata = TRUE)
  } else {
    input <- data
  }
  
  # check genotype column naming
  colnames(input) <- stringi::stri_replace_all_fixed(
    str = colnames(input), 
    pattern = "GENOTYPE", 
    replacement = "GT", 
    vectorize_all = FALSE
  )
  
  # necessary steps to make sure we work with unique markers and not duplicated LOCUS
  if (tibble::has_name(input, "LOCUS") && !tibble::has_name(input, "MARKERS")) {
    input <- dplyr::rename(.data = input, MARKERS = LOCUS)
  }
  
  # Compute count and Minor Allele Frequency -----------------------------------
  # We split the alleles here to prep for MAF
  # need to compute REF/ALT allele for non VCF file
  if (!tibble::has_name(input, "GT_VCF")) {
    ref.alt.alleles.change <- stackr::change_alleles(data = input)
    input <- dplyr::left_join(input, ref.alt.alleles.change$input, by = c("MARKERS", "INDIVIDUALS"))
  }
  
  # MAF
  betadiv <- dplyr::select(.data = input, MARKERS, POP_ID, GT_VCF) %>% 
    dplyr::filter(GT_VCF != "./.") %>%
    dplyr::group_by(MARKERS, POP_ID) %>%
    dplyr::summarise(
      N = as.numeric(n()),
      PQ = as.numeric(length(GT_VCF[GT_VCF == "1/0" | GT_VCF == "0/1"])),
      QQ = as.numeric(length(GT_VCF[GT_VCF == "1/1"]))
    ) %>%
    dplyr::mutate(MAF = ((QQ * 2) + PQ) / (2 * N)) %>%
    dplyr::ungroup(.) %>%
    dplyr::select(POP_ID, MARKERS, MAF) %>% 
    dplyr::group_by(POP_ID) %>% 
    tidyr::spread(data = ., key = MARKERS, value = MAF) %>%
    dplyr::ungroup(.) %>% 
    dplyr::mutate(POP_ID = as.integer(POP_ID))
  
  return(betadiv)
} # end write_betadiv

